src/parse.jl

```diff
@@ -68,20 +68,18 @@ const RULES = TupleDict(
 )
 
 # same as above, but valid for partially completed string commands. This is to determine when the key stack should be cleared.
-const PARTIAL_RULES = TupleDict(
-    # insert commands
-    r"^(?<c>[aAiIoO])" => InsertCommand,
-    # Special case: `0` is a motion command:
-    "^0\$" |> Regex => ZeroCommand,
-    # synonym commands
-    "^(?<n1>$REPEAT)(?<c>[xXCS])" |> Regex => SynonymCommand,
-    "^(?<n1>$REPEAT)($MOTION)" |> Regex => SimpleMotionCommand,
-    "^(?<n1>$REPEAT)((?|$(complex_motion())))" |> Regex => CompositeMotionCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(?<n2>$REPEAT)(?|($TEXTOBJECT)|($MOTION))" |> Regex => OperatorCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(?<n2>$REPEAT)((?|$(complex_motion())))" |> Regex => OperatorCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(\\k<op>)" |> Regex => LineOperatorCommand,
-    "^(?<n1>$REPEAT)r(.)" |> Regex => ReplaceCommand,
+const PARTIAL_RULES = (
+    "^(?<c>$INSERTCHARS)?\$" |> Regex,  # InsertCommand
+    "^0?\$" |> Regex,  # ZeroCommand
+    "^(?<n1>$REPEAT)(?<c>$DELETECHARS)?\$" |> Regex,  # SynonymCommand
+    "^(?<n1>$REPEAT)($MOTION)?\$" |> Regex,  # SimpleMotionCommand
+    "^(?<n1>$REPEAT)((?|$(complex_motion(true))))?\$" |> Regex,  # CompositeMotionCommand
+    "^(?<n1>$REPEAT)((?<op>$OPERATOR)((?<n2>$REPEAT)((?|($PARTIALTEXTOBJECT)|($MOTION)))?)?)?\$" |> Regex,  # OperatorCommand
+    "^(?<n1>$REPEAT)((?<op>$OPERATOR)((?<n2>$REPEAT)((?|$(complex_motion(true))))?)?)?\$" |> Regex,  # OperatorCommand (2)
+    "^(?<n1>$REPEAT)((?<op>$OPERATOR)(\\k<op>)?)?\$" |> Regex,  # LineOperatorCommand
+    "^(?<n1>$REPEAT)(r(.)?)?\$" |> Regex  # ReplaceCommand
 )
```

Comment:
These are essentially `RULES`, but with the pattern:
```
^(a(b(c)?)?)?$
```
which allows for `a`, `ab`, and `abc` to be matched, but not `b` or `c` or `bc`.